src/sparse/KokkosSparse_sptrsv_supernode.hpp

```diff
@@ -1109,46 +1170,105 @@ invert_supernodal_columns(KernelHandle kernelHandle, bool unit_diag, int nsuper,
+  double time3 = 0.0;
+  #endif
 
   // ----------------------------------------------------------
   // now let's invert some blocks
+  // > first go through all the supernode columns
+  // > use KokkosBlas on large blocks, and keep track of small blocks
+  // > to call batchedBlas on them
+  int num_batchs = 0;
+  int size_unblocked = handle->get_supernode_size_unblocked();
+  integer_view_host_t supernode_ids ("supernode_batch", nsuper);
   for (int s2 = 0; s2 < nsuper; s2++) {
-    int j1 = nb[s2];
-    int nsrow = hr(j1+1) - hr(j1);
     int nscol = nb[s2+1] - nb[s2];
 
-    auto nnzD = hr (j1);
-    char uplo_char = (lower ? 'L' : 'U');
-    char diag_char = (unit_diag ? 'U' : 'N');
+    if (nscol >= size_unblocked) {
+      int j1 = nb[s2];
+      int nsrow = hr(j1+1) - hr(j1);
 
-    Kokkos::View<scalar_t**, Kokkos::LayoutLeft, memory_space, Kokkos::MemoryUnmanaged>
-      viewL (&hv(nnzD), nsrow, nscol);
-    auto Ljj = Kokkos::subview (viewL, range_type (0, nscol), Kokkos::ALL ());
+      auto nnzD = hr (j1);
+      char uplo_char = (lower ? 'L' : 'U');
+      char diag_char = (unit_diag ? 'U' : 'N');
 
-    timer.reset ();
-    KokkosBlas::trtri(&uplo_char, &diag_char, Ljj);
-    time1 += timer.seconds ();
-
-    if (nsrow > nscol && invert_offdiag) {
-      char side_char = 'R';
-      char tran_char = 'N';
-      auto Lij = Kokkos::subview (viewL, range_type (nscol, nsrow), Kokkos::ALL ());
+      Kokkos::View<scalar_t**, Kokkos::LayoutLeft, memory_space, Kokkos::MemoryUnmanaged>
+        viewL (&hv(nnzD), nsrow, nscol);
+      auto Ljj = Kokkos::subview (viewL, range_type (0, nscol), Kokkos::ALL ());
 
+      #ifdef KOKKOS_SPTRSV_SUPERNODE_PROFILE
       timer.reset ();
-      KokkosBlas::trmm (&side_char, &uplo_char,
-                        &tran_char, &diag_char,
-                        one, Ljj, Lij);
-      time2 += timer.seconds ();
+      #endif
+      KokkosBlas::trtri(&uplo_char, &diag_char, Ljj);
+      #ifdef KOKKOS_SPTRSV_SUPERNODE_PROFILE
+      time1 += timer.seconds ();
+      #endif
+
+      if (nsrow > nscol && invert_offdiag) {
+        char side_char = 'R';
+        char tran_char = 'N';
+        auto Lij = Kokkos::subview (viewL, range_type (nscol, nsrow), Kokkos::ALL ());
+
+        #ifdef KOKKOS_SPTRSV_SUPERNODE_PROFILE
+        timer.reset ();
+        #endif
+        KokkosBlas::trmm (&side_char, &uplo_char,
+                          &tran_char, &diag_char,
+                          one, Ljj, Lij);
+        #ifdef KOKKOS_SPTRSV_SUPERNODE_PROFILE
+        time2 += timer.seconds ();
+        #endif
+      }
     }
+    else {
+      supernode_ids (num_batchs) = s2;
+      num_batchs ++;
+    }
+  }
+  // now call batchedBLAS
+  if (num_batchs > 0) {
+    using range_policy = Kokkos::RangePolicy<execution_space>;
+    using Uplo = KokkosBatched::Uplo;
+    using Diag = KokkosBatched::Diag;
+    #ifdef KOKKOS_SPTRSV_SUPERNODE_PROFILE
+    timer.reset ();
+    #endif
+    if (lower) {
+      if (unit_diag) {
```

Comment:
Same as above. Consider using a single `TriSupernodeTrtriFunctor` invocation by moving these conditionals to https://github.com/kokkos/kokkos-kernels/pull/797/files#diff-3205ca442d81699d8344ea8328d91f81R1172 and setting `using uplo_arg = Uplo::{Lower,Upper}` and `using diag_arg = Diag::{Unit,NonUnit}`.

Reply:
@iyamazaki, @ndellingwood  To support a single `TriSupernodeTrtriFunctor` invocation the sptrsv handle and callers would need to be refactored such that [lower](https://github.com/kokkos/kokkos-kernels/blob/ad1c5d0ffd7b73bbc836d610a6acb96ddcc7cc88/src/sparse/KokkosSparse_sptrsv_handle.hpp#L247), [unit_diag](https://github.com/kokkos/kokkos-kernels/blob/ad1c5d0ffd7b73bbc836d610a6acb96ddcc7cc88/src/sparse/KokkosSparse_sptrsv_handle.hpp#L319), and [invert_off_diag](https://github.com/kokkos/kokkos-kernels/blob/ad1c5d0ffd7b73bbc836d610a6acb96ddcc7cc88/src/sparse/KokkosSparse_sptrsv_handle.hpp#L347) are template parameters. This seems like a significant effort but could improve readability and maintainability in the long run; is it worth creating a issue with a cleanup label for this?

Reply:
@e10harvey good point about cleanup, would you be okay that we open a new issue and focus on that in a separate effort for this PR?

Reply:
@ndellingwood: Yes, I'll create an issue.

Reply:
Thank you!! So, any additional changes are needed for this PR, or will be addressed by the next one, @e10harvey ?

Reply:
@iyamazaki I think this PR just needs spot-checks run. I'll run and post those later today.

Reply:
Thank you, @e10harvey !! Do you have some details for the failed build?

Reply:
> Thank you, @e10harvey !! Do you have some details for the failed build?

No problem. I do and I'm testing the fix now. If it passes, I'll post PR with that so we can get this merged.